typescript/sdk/src/ism/EvmIsmModule.ts

```diff
@@ -1,59 +1,691 @@
-import { Address, ProtocolType, rootLogger } from '@hyperlane-xyz/utils';
+import { ethers } from 'ethers';
+import { Logger } from 'pino';
 
-import { HyperlaneAddresses } from '../contracts/types.js';
+import {
+  DefaultFallbackRoutingIsm__factory,
+  DomainRoutingIsm,
+  DomainRoutingIsmFactory__factory,
+  DomainRoutingIsm__factory,
+  IAggregationIsm,
+  IAggregationIsm__factory,
+  IInterchainSecurityModule__factory,
+  IMultisigIsm,
+  IMultisigIsm__factory,
+  IRoutingIsm,
+  MailboxClient__factory,
+  OPStackIsm__factory,
+  Ownable__factory,
+  PausableIsm__factory,
+  StaticAddressSetFactory,
+  StaticThresholdAddressSetFactory,
+  TestIsm__factory,
+  TrustedRelayerIsm__factory,
+} from '@hyperlane-xyz/core';
+import {
+  Address,
+  ProtocolType,
+  assert,
+  configDeepEquals,
+  eqAddress,
+  normalizeConfig,
+  objFilter,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import { attachAndConnectContracts } from '../contracts/contracts.js';
+import { HyperlaneAddresses, HyperlaneContracts } from '../contracts/types.js';
 import {
   HyperlaneModule,
   HyperlaneModuleArgs,
 } from '../core/AbstractHyperlaneModule.js';
 import { HyperlaneDeployer } from '../deploy/HyperlaneDeployer.js';
-import { ProxyFactoryFactories } from '../deploy/contracts.js';
+import {
+  ProxyFactoryFactories,
+  proxyFactoryFactories,
+} from '../deploy/contracts.js';
+import { extractOwnerAddress } from '../deploy/types.js';
 import { MultiProvider } from '../providers/MultiProvider.js';
-import { EthersV5Transaction } from '../providers/ProviderType.js';
-import { ChainNameOrId } from '../types.js';
+import {
+  AnnotatedEthersV5Transaction,
+  createAnnotatedEthersV5Transaction,
+} from '../providers/ProviderType.js';
+import { ChainMap, ChainName, ChainNameOrId } from '../types.js';
+import { findMatchingLogEvents } from '../utils/logUtils.js';
 
 import { EvmIsmReader } from './EvmIsmReader.js';
-import { IsmConfig } from './types.js';
+import {
+  AggregationIsmConfig,
+  DeployedIsm,
+  IsmConfig,
+  IsmType,
+  MultisigIsmConfig,
+  MutableIsmType,
+  PausableIsmConfig,
+  RoutingIsmConfig,
+} from './types.js';
+import { calculateDomainRoutingDelta } from './utils.js';
+
+type IsmModuleAddresses = {
+  deployedIsm: Address;
+  mailbox: Address;
+};
 
 export class EvmIsmModule extends HyperlaneModule<
   ProtocolType.Ethereum,
   IsmConfig,
-  HyperlaneAddresses<ProxyFactoryFactories> & {
-    deployedIsm: Address;
-  }
+  HyperlaneAddresses<ProxyFactoryFactories> & IsmModuleAddresses
 > {
-  protected logger = rootLogger.child({ module: 'EvmIsmModule' });
-  protected reader: EvmIsmReader;
+  protected readonly logger = rootLogger.child({ module: 'EvmIsmModule' });
+  protected readonly reader: EvmIsmReader;
+  protected readonly factories: HyperlaneContracts<ProxyFactoryFactories>;
+
+  // Adding these to reduce how often we need to grab from MultiProvider.
+  public readonly chainName: string;
```

Comment:
nit: use `chain: ChainName` to be consistent?

Reply:
also updated `domainId` below to `Domain`